scientific-computing-with-python/learn-inheritance-by-building-an-equation-solver/quadratic_equation_solver.py

```diff
@@ -0,0 +1,145 @@
+from abc import ABC, abstractmethod
+import re
+
+
+class Equation(ABC):
+    """
+    Equation objects are instantiated providing the equation coefficients as arguments.
+    The coefficient order goes from the zero-th grade of x to the n-th grade.
+    Coefficients with the value of zero must be specified.
+    """
+    def __init__(self, *args):
+        if (self.grade + 1)!= len(args):
+            raise TypeError(f"'Equation' object takes {self.grade + 1} positional arguments but {len(args)} were given")
+        for arg in args:            
+            if not isinstance(arg, (int, float)):
+                raise TypeError("Coefficients must be of type 'int' or 'float'")
+        if args[-1] == 0:
+            raise ValueError('Highest order coefficient must be different from zero')
+        self.coefficients = {n: arg for n, arg in enumerate(args)}
+        self.results = []
+        self.details = []        
+
+    def __str__(self):
+        terms = []        
+        for n, coefficient in self.coefficients.items():            
+            if coefficient:
+                if n == 0:
+                    terms.append(f'{coefficient:+}')
+                elif n == 1:
+                    terms.insert(0, f'{coefficient:+}x')
+                else:
+                    terms.insert(0, f'{coefficient:+}x**{n}')
+        equation_string = ' '.join(terms)
+        equation_string += ' = 0'
+        return re.sub(r'(?<!\d)1(?=x)', '', equation_string.strip('+'))
```

Comment:
To improve readability, I would suggest to refactor the `__str__` method.

```suggestion
   def __str__(self):
        return self.format_equation()

    def format_equation(self):
        terms = []
        for n, coefficient in self.coefficients.items():
            if coefficient:
                if n == 0:
                    terms.append(f'{coefficient:+}')
                elif n == 1:
                    terms.insert(0, f'{coefficient:+}x')
                else:
                    terms.insert(0, f'{coefficient:+}x**{n}')
        equation_string = ' '.join(terms)
        equation_string += ' = 0'
        return re.sub(r'(?<!\d)1(?=x)', '', equation_string.strip('+'))
```